rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+  for i in range(total_steps):
+    strategy.schedule(replica_fn, args=(distributed_iter,))
+# … omitted
+```
+
+
+**We will start with this kind of training loop in our first version.**
+
+
+### Fault Tolerance
+
+
+#### Task Failure
+
+
+##### Worker failure
+
+
+###### When scheduling
+
+When a worker fails, our training will continue without this failed worker. Functions scheduled on a failed worker will be rescheduled on other workers. 
+
+For functions that bound to a specific worker, e.g. resource creation function, they will be queued until the worker is back.
+
+When the failed worker is back, we will update the cluster configuration with `context.update_server_def` which would also reset all the states. After resources on the restarted worker are built, we can resume scheduling functions on the worker. 
+
+
+###### When materialing a `Future` object
+
+It is possible that a function is executed but its corresponding worker fails when users try to consume its output. In this case, we will give users a `None` value and set an error in the `Future` object.
+
+We can mitigate the problem by eagerly materializing function outputs when they are passed to `local_results`. 
+
+We can explore mechanisms to recover these objects in the future. In the short-term, users can choose to write the results to variables on parameter servers, just like a Keras metric.
+
+
+##### Parameter server failure
+
+When a parameter server fails, the error will be propagated to the client via workers. Since the latest values of variables on the failed parameter servers are gone, there is no way for the client to recover them. Therefore the training will pause until the failed parameter server is back. The client then needs to clean up other variables on other parameter servers, rebuild all the variables and load variable values from a checkpoint. To trigger this process, the simplest method is to restart the client as well.
+
+
+##### Client failure
+
+When a client fails, some scheduled functions will continue to run on workers. No new functions will be scheduled. When the client comes back, it will create variables, load from a checkpoint, schedule functions with a new context id. All the old variables will be garbage-collected when we reset their eager contexts.
+
+#### Resource Management for Workers
+
+When a worker has recovered from failure, we will need to rebuild iterators, worker-local variables, lookup tables and other resources on that worker that don’t need to be read from a checkpoint. This means that the client will have to keep track of these iterators, worker-local variables and other resources.
+
+Keeping track of resources and rebuilding them will be achieved depending how users create their resources:
+
+
+*   we will record iterators created via `tf.distribute`’s API; The state of a rebuilt iterator will be lost. We can recover their states as future work.
+*   we will capture the creation of worker-local variables via variable creator scopes.
```

Comment:
Who owns the var creator scopes? Where is the state and memory held in the event of worker failure?

Reply:
All the variable creation will be captured by distribution strategies' scope. Worker-local variables will be created on workers, the client just holds references to these variables, similar to an iterator.